JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,536 @@
+  onClick={this.onClickHandler}
+>
+  <InnerChildComponent className={styles.innerChild} />
+</OuterChildComponent>
+```
+
+- always wrap *render* method's return statement in parentheses (it's optional if it's returning a single line of JSX)
+
+``` jsx
+// bad 
+render() {
+  // ...
+  return <div>
+      <ChildComponent />
+    </div>;
+}
+
+// bad 
+render() {
+  // ...
+  return
+    <div>
+      <ChildComponent />
+    </div>;
+}
+
+// good
+render() {
+  // ...
+  return (
+    <div>
+      <ChildComponent />
+    </div>
+  );
+}
+
+// good
+render() {
+  // ...
+  return <ChildComponent />;
+}
+
+// good
+render() {
+  // ...
+  return (
+    <ChildComponent />
+  );
+}
+```
+
+### Passing Props
+Use camelCase for prop names and double quotes for JSX attribute values if you are passing them as a *string*.
+
+``` jsx
+class ComponentName extends React.Component {
+  // ...
+  render () {
+    return (
+      <ChildComponent
+        defaultTitle="Some Title"
+        title={this.props.title}
+      />
+    );
+  }
+}
+```
+
+Declare functions before passing them as props instead of passing inline anonymous (arrow) functions. They will be needlessy re-rendered each time, thus effecting performance.
+
+``` jsx
+// bad
+<ChildComponent onClick={() => { /* ... */ } />
+    
+// good
+class ComponentName extends React.Component {
+  onClickHandler() {
+    // ...
+  }
+
+  render () {
+    return (
+      <ChildComponent onClick={this.onClickHandler} />
+    );
+  }
+}
+```
+
+Omit values for boolean props if default value is **true**.
+
+``` jsx
+// bad
+<ChildComponent
+  onClick={this.onClickHandler}
+  isPropActive={true}
+/>
+
+// good
+<ChildComponent
+  onClick={this.onClickHandler}
+  isPropActive
+/>
+```
+
+### Receiving Props
+*Checkout [`prop-types`](https://github.com/facebook/prop-types) lib on GitHub for general info.*
+
+In class components:
+- always use prop validation
+- use **static member** for declaring prop types and default props (if your transpiler doesn't support static members, set them after class declaration and then export the class; See: ["Exporting Components" section](#exporting-components))
+- use *defaultProps* if you need to set default values
+
+``` jsx
+export default class ComponentName extends React.Component {
+  static propTypes = {
+    id: PropTypes.string.isRequired,
+    name: PropTypes.string,
+  };
+
+  static defaultProps = {
+    name: 'John Doe',
+  };
+
+  // ...
+
+  render() {
+    const {name} = this.props;
+
+    return (
+      <div>{name}</div>
+    );
+  }
+}
+```
+
+In stateless components, use PropTypes for type checking but use native JS for setting default values.
+
+``` jsx
+function ComponentName({id, name = 'John Doe'}) {
+  // ...
+};
+
+ComponentName.propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string
+}
+```
+
+### Using props
+
+If you are using more than one property in a method, preffer accessing them with **object destructuring**.
```

Comment:
Mention also extracting complex variables in scope instead of defining them in JSX

```
render() {
    const trasnlatedLabel = label || this.t('translation');
    
    return (
           <Component label={trasnlatedLabel} />
    );
}
```